internal/overlord/logstate/loki.go

```diff
@@ -0,0 +1,162 @@
+// Copyright (c) 2022 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package logstate
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/canonical/pebble/internal/logger"
+	"github.com/canonical/pebble/internal/servicelog"
+)
+
+const (
+	requestStart = `
+{
+  "streams": [
+    {
+      "stream": {
+        "service": %q
+      },
+      "values": [`
+
+	requestEnd = `
+      ]
+    }
+  ]
+}`
+
+	lokiPushAPI = "/loki/api/v1/push"
+)
+
+// lokiClient implements logClient
+// We send a new request to Loki when either:
+//   - we have maxMessages ready to send; or
+//   - delay has passed since the last request.
+type lokiClient struct {
+	service     string
+	lokiURL     *url.URL
+	maxMessages int
+	delay       time.Duration
+
+	// buffered request so far
+	buffer *bytes.Buffer
+	// how many messages have been written so far
+	messages int
+	// write deadline for flushing buffer
+	deadline *time.Time
+}
+
+func newLokiClient(service, url string, maxMessages int, delay time.Duration) (*lokiClient, error) {
```

Comment:
At some point we have to parse the URL from a string into a URL object. I'm not sure where the best place to do that is

Reply:
I disagree that things like `maxMessages` should be unsigned. Go stdlib API's tend to use just `int`, and the reason is because otherwise 0 -- a very common value -- is a very problematic wraparound point. If the caller accidentally provides `x-1` and `x` is right on the boundary (i.e., 0) then you'll get a huge negative number, which is almost certainly not what you want. Whereas with `int` the wraparound occurs at -9223372036854775808 or +9223372036854775807, which is miles away from 0. Ian Lance Taylor [explains](https://groups.google.com/g/golang-nuts/c/F181gqrN0sQ/m/hQp3gQpqCQAJ) it better than I can:

> > From what I've read in the past, unsigned values are generally not
> > used because of the risk of accidental overflow bugs.
> 
> Yes, that is the reason.
> 
> Computer integers inevitably have surprising behavior for certain
> values. For signed integers those values are very large and very
> small. For unsigned integers, that value is zero. Zero is far more
> likely to occur in ordinary use than very large or very small numbers.
> If you are doing any kind of arithmetic, it's better to use signed
> integer types. You are less likely to have surprising results. 

Regarding using the `URL` type, I kinda disagree there too. That tends to be used when you're parsing/splitting URLs, not for passing in an API URL. For example, the stdlib `Get` and `Post` methods take `url string`, not a URL object. It makes them simpler to use and work with.

Reply:
Yep I can agree on the unsigned part. Old C++ habits die hard. 

As for the URL my reasoning is this. If we were dealing with the raw value and not doing any operations on it a string would be fine. But we are doing work on the value to add paths to get the final destination.

However happy to go either way and would rather convention that is inline with the rest of world.

Reply:
if we're not adding defaults we can use a plain string for URL